app/src/main/java/com/pileproject/drive/execution/RxObservableProgram.java

```diff
@@ -0,0 +1,190 @@
+import rx.Subscriber;
+
+/**
+ * A concrete class of {@link rx.Observable.OnSubscribe}, which generates
+ * a stream of blocks that was executed during a program execution.
+ * <p/>
+ * The stream emits {@link Bundle} objects for the subscriber.
+ * The bundles contain message type whose key is {@link RxObservableProgram#KEY_MESSAGE_TYPE},
+ * and optionally one message argument whose key is {@link RxObservableProgram#KEY_MESSAGE_ARG}.
+ * The message types are:
+ * <ul>
+ *     <li>{@link RxObservableProgram#START}: emitted at the beginning of the execution</li>
+ *     <li>{@link RxObservableProgram#END}: emitted at the end of the execution</li>
+ *     <li>{@link RxObservableProgram#PAUSE}: emitted at {@link RxObservableProgram#requestPause()} is called</li>
+ *     <li>{@link RxObservableProgram#BLOCK_INDEX}: emitted when a block was executed.
+ *         this message contains an argument which is the index of the block </li>
+ * </ul>
+ * <p/>
+ * Typically you can use this class with code like below.
+ * Note that the process of this class is heavy, including I/O connection.
+ * Do not run this on your UI thread.
+ *
+ * <code>
+ *     RxObservableProgram program = new RxObservableProgram(...);
+ *     Observable.create(program)
+ *         .subscribeOn(Schedulers.newThread())
+ *         .subscribe(...);
+ * </code>
+ */
+public class RxObservableProgram implements Observable.OnSubscribe<Bundle> {
+
+    public static final String KEY_MESSAGE_TYPE = "key";
+    public static final String KEY_MESSAGE_ARG = "arg";
+    public static final int START = 1;
+    public static final int BLOCK_INDEX = 3;
+    public static final int PAUSE = 4;
+    public static final int END = 2;
+
+    private final MachineController controller;
+
+    private final ExecutionCondition condition;
+
+    private boolean terminateRequest;
+
+    private boolean pauseRequest;
+
+    private Thread runningThread;
+
+    /**
+     * @param program list of blocks to be executed.
+     * @param controller controller for the machine to be manipulated.
+     */
+    public RxObservableProgram(List<BlockBase> program, MachineController controller) {
+
+        this.controller = controller;
+        this.condition = new ExecutionCondition(program);
+    }
+
+    @Override
+    public void call(Subscriber<? super Bundle> subscriber) {
+
+        runningThread = Thread.currentThread();
+        subscriber.onNext(makeMessageBundle(START));
+
+        try {
+            mainLoop(subscriber);
+        } catch (RuntimeException e) {
+            subscriber.onError(e);
+        } finally {
+            controller.close();
+        }
+
+        subscriber.onCompleted();
+    }
+
+    private void mainLoop(Subscriber<? super Bundle> subscriber) throws RuntimeException {
+
+        boolean haltAlreadyCalled = false;
+
+        while (!condition.hasProgramFinished()) {
+
+            if (terminateRequest) {
+                subscriber.onNext(makeMessageBundle(END));
```

Comment:
Okay, it's reasonable but I think `END` is somewhat confusing because the meaning of it looks the same as `subscriber.onCompleted()`. I propose some options for this;

1. Remove `END` from this class (`onCompleted()` does the same thing)
2. Rename `END` to `TERMINATE_BY_USER` or something (`onCompleted()` notifies the normal ending)
3. Keep using this `END` but remove it from here and insert it after this while loop so that this `END` express all the end cases